### PR TITLE
Define the task manager interface and data model

### PR DESCRIPTION
**Pull request by ywk253100**
Define the task manager interface and data model

Signed-off-by: Wenkai Yin <yinw@vmware.com>

### CODE REVIEW

File: make/migrations/postgresql/0040_2.1.0_schema.up.sql [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+CREATE TABLE IF NOT EXISTS task (
+    id SERIAL PRIMARY KEY NOT NULL,
+    execution_id int NOT NULL,
+    job_id varchar(64),
+    status varchar(16) NOT NULL,
+    status_code int NOT NULL,
+    status_revision int,
+    status_message text,
+    retry_count int,
```
Comment (steven-zou) [inline make/migrations/postgresql/0040_2.1.0_schema.up.sql:21]: What does `retry_count` use?

File: make/migrations/postgresql/0040_2.1.0_schema.up.sql [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    extra_attrs JSONB,
+    start_time timestamp DEFAULT CURRENT_TIMESTAMP,
+    update_time timestamp DEFAULT CURRENT_TIMESTAMP,
+    end_time timestamp NULL,
+    FOREIGN KEY (execution_id) REFERENCES execution(id)
+);
+
+CREATE TABLE IF NOT EXISTS check_in_data (
```
Comment (steven-zou) [inline make/migrations/postgresql/0040_2.1.0_schema.up.sql:29]: I'm thinking, do we really need to persist check-in data? Most of the time, the business component should process check-in data when it's transformed from JS. Maybe passing it to the data processors is enough?

File: src/pkg/task/manager.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+// Manager provides an execution-task model to abstract the interactive with jobservice.
+// All of the operations with jobservice should be delegated by this manager
+type Manager interface {
+	// CreateExecution creates an execution. The "typee" is used to specify the type(execution, scan, etc.)
+	// of the execution, and the "extraAttrs" can be used to set the customized attributes
+	CreateExecution(ctx context.Context, typee, trigger string, extraAttrs ...map[string]interface{}) (id int64, err error)
+	// UpdateExecutionStatus updates the status of the execution.
+	// In most cases, the execution status can be calculated from the referenced tasks automatically.
+	// When the execution contains no tasks or failed to create tasks, the status should be set manually
+	UpdateExecutionStatus(ctx context.Context, id int64, status, message string) (err error)
```
Comment (steven-zou) [inline src/pkg/task/manager.go:32]: Support both options. For simplification, maybe we only need to support one. Calculating the data from the sub-tasks?

File: src/pkg/task/model.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	ExecutionTypeRetention         = "RETENTION"
+	ExecutionTypeScan              = "SCAN"
+	ExecutionTypeScanAll           = "SCAN_ALL"
+	ExecutionTypeScheduler         = "SCHEDULER"
+
+	ExecutionTriggerManual   = "MANUAL"
+	ExecutionTriggerSchedule = "SCHEDULE"
+	ExecutionTriggerEvent    = "EVENT"
+
+	StatusSucceeded = "SUCCEEDED"
```
Comment (steven-zou) [inline src/pkg/task/model.go:37]: Can we directly use `job.Status` at this time?

File: src/pkg/task/model.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	ID int64 `json:"id"`
+	// indicate the execution type: replication/GC/retention/scan/etc.
+	Type   string `json:"type"`
+	Status string `json:"status"`
+	// the detail message to explain the status in some cases. e.g.
+	// 1. After creating the execution, there may be some errors before creating tasks, the
+	// "StatusMessage" can contain the error message
+	// 2. The execution may contain no tasks, "StatusMessage" can be used to explain the case
+	StatusMessage      string `json:"status_message"`
+	TaskCount          int64  `json:"task_count"`
```
Comment (steven-zou) [inline src/pkg/task/model.go:56]:
```
TaskCount          int64  `json:"task_count"`
	SucceededTaskCount int64  `json:"succeeded_task_count"`
	FailedTaskCount    int64  `json:"failed_task_count"`
	PendingTaskCount   int64  `json:"pending_task_count"`
	RunningTaskCount   int64  `json:"running_task_count"`
	ScheduledTaskCount int64  `json:"scheduled_task_count"`
	StoppedTaskCount   int64  `json:"stopped_task_count"`
```

How about defining a `Metrics` struct to contain such data?

```
type Execution struct {
// ...
metrics *Metrics
// ...

}
```

File: make/migrations/postgresql/0040_2.1.0_schema.up.sql (hunk 1 above)
Comment (ywk253100) [reply make/migrations/postgresql/0040_2.1.0_schema.up.sql:21]: Some jobs may retry, use this field to record the retry times

File: src/pkg/task/manager.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+	// DeleteExecution deletes the specified execution and its tasks
+	DeleteExecution(ctx context.Context, id int64) (err error)
+	// GetExecution gets the specified execution
+	GetExecution(ctx context.Context, id int64) (execution *Execution, err error)
+	// ListExecutions lists executions according to the query
+	ListExecutions(ctx context.Context, query *q.Query) (executions []*Execution, err error)
+	// CreateTask submits the job to jobservice and creates a corresponding task record.
+	// An execution must be created first and the task will be linked to it.
+	// The "extraAttrs" can be used to set the customized attributes
+	CreateTask(ctx context.Context, executionID int64, job *Job, extraAttrs ...map[string]interface{}) (id int64, err error)
```
Comment (steven-zou) [inline src/pkg/task/manager.go:44]: Can we split this interface into two interfaces? One is for execution and one is for the task.

File: src/pkg/task/model.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+	VendorType string `json:"vendor_type"`
+	// the ID of vendor policy/rule/etc. e.g. replication policy ID
+	VendorID int64  `json:"vendor_id"`
+	Status   string `json:"status"`
+	// the detail message to explain the status in some cases. e.g.
+	// 1. After creating the execution, there may be some errors before creating tasks, the
+	// "StatusMessage" can contain the error message
+	// 2. The execution may contain no tasks, "StatusMessage" can be used to explain the case
+	StatusMessage string   `json:"status_message"`
+	Metrics       *Metrics `json:"meetrics"`
```
Comment (bitsf) [inline src/pkg/task/model.go:50]: meetrics typo

**Pull request merged**